File: agent_arena/agents/agent.py
"""Agent class using raw Anthropic API for reliable web server compatibility."""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, List
import uuid
import asyncio
import logging
import os

import anthropic

from ..core.types import AgentStatus, AgentConfig
from ..core.message import Message

logger = logging.getLogger(__name__)


@dataclass
class Agent:
    """
    An AI agent in the arena using raw Anthropic API.

    Each agent has:
    - Its own conversation history
    - Custom personality via system prompt
    - Manual step control
    """

    name: str
    description: str = ""
    system_prompt: str = ""

    # Personality
    personality_traits: dict[str, float] = field(default_factory=dict)
    speaking_style: str = ""
    interests: List[str] = field(default_factory=list)

    # Behavior
    response_tendency: float = 0.5
    temperature: float = 0.7
    model: str = "claude-haiku-4-5-20251001"
    tools: List[str] = field(default_factory=list)

    # Identity
    id: str = field(default_factory=lambda: str(uuid.uuid4()))

    # Runtime state
    status: AgentStatus = AgentStatus.OFFLINE
    last_spoke_at: Optional[datetime] = None
    message_count: int = 0

    # Conversation history for this agent
    conversation_history: List[dict] = field(default_factory=list)

    # Anthropic client
    _client: Optional[anthropic.AsyncAnthropic] = field(default=None, repr=False)

    def __post_init__(self):
        """Build the full system prompt."""
        if not self.system_prompt:
            self.system_prompt = self._build_default_prompt()

    def _build_default_prompt(self) -> str:
        """Build a default system prompt from personality traits."""
        traits_str = ", ".join(
            f"{k}: {v:.1f}" for k, v in self.personality_traits.items()
        ) if self.personality_traits else "balanced"

        interests_str = ", ".join(self.interests) if self.interests else "general topics"

        return f"""You are {self.name}, an AI participant in a group chat with other AI agents and humans.

{self.description}

Your personality traits: {traits_str}
Your speaking style: {self.speaking_style or 'natural and conversational'}
Your interests: {interests_str}

IMPORTANT RULES:
1. Stay in character as {self.name}
2. To address another participant, use @TheirName
3. Keep responses concise (1-3 sentences typically, like IRC chat)
4. You can disagree, ask questions, build on ideas, or change topics
5. Don't repeat yourself or others unnecessarily
6. If you truly have nothing meaningful to add, respond with just "[PASS]"
7. Be authentic to your personality - don't just agree with everyone
8. Just write your response directly - no timestamps or formatting
"""

    async def connect(self) -> None:
        """Initialize the Anthropic client."""
        api_key = os.environ.get("ANTHROPIC_API_KEY")
        if not api_key:
            raise ValueError("ANTHROPIC_API_KEY environment variable not set")

        self._client = anthropic.AsyncAnthropic(api_key=api_key)
        self.status = AgentStatus.IDLE
        self.conversation_history = []
        logger.info(f"Agent {self.name} connected")

    async def disconnect(self) -> None:
        """Disconnect the agent."""
        self._client = None
        self.status = AgentStatus.OFFLINE
        self.conversation_history = []
        logger.info(f"Agent {self.name} disconnected")

    async def step(self, chat_context: str) -> Optional[str]:
        """
        Execute one step - generate a response to the current chat context.

        This is the manual step function for individual agent control.
        """
        if not self._client:
            raise RuntimeError(f"Agent {self.name} is not connected")

        self.status = AgentStatus.THINKING
        logger.info(f"Agent {self.name} thinking...")

        try:
            # Build the message
            user_message = f"""Current conversation:

{chat_context}

Now respond as {self.name}. Keep it brief (1-2 sentences). Just your response, nothing else."""

            # Add to conversation history
            self.conversation_history.append({
                "role": "user",
                "content": user_message
            })

            # Keep history manageable (last 20 exchanges)
            if len(self.conversation_history) > 40:
                self.conversation_history = self.conversation_history[-40:]

            # Call Anthropic API
            response = await self._client.messages.create(
                model=self.model,
                max_tokens=10000,
                system=self.system_prompt,
                messages=self.conversation_history,
                temperature=self.temperature
            )

            # Extract response text
            response_text = response.content[0].text.strip()

            # Add assistant response to history
            self.conversation_history.append({
                "role": "assistant",
                "content": response_text
            })

            # Check for [PASS]
            if response_text == "[PASS]":
                self.status = AgentStatus.IDLE
                return None

            self.status = AgentStatus.SPEAKING
            self.last_spoke_at = datetime.now()
            self.message_count += 1

            logger.info(f"Agent {self.name} responded: {response_text[:50]}...")
            return response_text

        except Exception as e:
            logger.error(f"Agent {self.name} error: {e}")
            self.status = AgentStatus.IDLE
            return None
        finally:
            self.status = AgentStatus.IDLE

    # Alias for compatibility
    async def respond(self, context: str, timeout: float = 60.0) -> Optional[str]:
        """Generate a response (alias for step)."""
        return await self.step(context)

    def should_respond(self, message: Message, all_agents: List[str]) -> float:
        """Calculate probability that this agent should respond."""
        base_prob = self.response_tendency * 0.3

        # Direct mention - high priority
        if self.name.lower() in [m.lower() for m in message.mentions]:
            return min(0.95, base_prob + 0.6)

        # Question - medium boost
        if "?" in message.content:
            base_prob += 0.15

        # Topic matches interests
        content_lower = message.content.lower()
        for interest in self.interests:
            if interest.lower() in content_lower:
                base_prob += 0.1
                break

        # Recently spoke - reduce
        if self.last_spoke_at:
            seconds_ago = (datetime.now() - self.last_spoke_at).total_seconds()
            if seconds_ago < 10:
                base_prob *= 0.3
            elif seconds_ago < 30:
                base_prob *= 0.6

        return min(0.8, base_prob)

    @classmethod
    def from_config(cls, config: AgentConfig) -> "Agent":
        """Create an agent from a config dict."""
        # Map short model names to full IDs
        model_map = {
            "haiku": "claude-haiku-4-5-20251001",
            "sonnet": "claude-sonnet-4-20250514",
            "opus": "claude-opus-4-20250514",
        }
        model = config.get("model", "claude-haiku-4-5-20251001")
        model = model_map.get(model, model)

        return cls(
            id=config.get("id", str(uuid.uuid4())),
            name=config["name"],
            description=config.get("description", ""),
            system_prompt=config.get("system_prompt", ""),
            personality_traits=config.get("personality_traits", {}),
            speaking_style=config.get("speaking_style", ""),
            interests=config.get("interests", []),
            response_tendency=config.get("response_tendency", 0.5),
            temperature=config.get("temperature", 0.7),
            model=model,
            tools=config.get("tools", []),
        )

    def to_dict(self) -> dict:
        """Serialize agent to dictionary."""
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "personality_traits": self.personality_traits,
            "speaking_style": self.speaking_style,
            "interests": self.interests,
            "response_tendency": self.response_tendency,
            "temperature": self.temperature,
            "model": self.model,
            "status": self.status.value,
            "message_count": self.message_count,
            "last_spoke_at": self.last_spoke_at.isoformat() if self.last_spoke_at else None,
        }

File: src/agents/agent.ts
/**
 * Agent class using Anthropic SDK for Node.js.
 * Now with SQLite persistence and tool support.
 */

import Anthropic from '@anthropic-ai/sdk';
import { v4 as uuidv4 } from 'uuid';
import { AgentStatus, AgentConfig, PersonalityTraits, AgentData } from '../core/types.js';
import { Message } from '../core/message.js';
import * as db from '../core/database.js';
import {
  ToolContext,
  getAllToolDefinitions,
  executeTool,
  getAgentWorkspace,
  getSharedWorkspace
} from '../tools/index.js';

// Model name mapping
const MODEL_MAP: Record<string, string> = {
  haiku: 'claude-haiku-4-5-20251001',
  sonnet: 'claude-sonnet-4-20250514',
  opus: 'claude-opus-4-20250514'
};

export class Agent {
  id: string;
  name: string;
  description: string;
  system_prompt: string;
  personality_traits: PersonalityTraits;
  speaking_style: string;
  interests: string[];
  response_tendency: number;
  temperature: number;
  model: string;
  tools: string[];
  status: AgentStatus;
  last_spoke_at: Date | null;
  message_count: number;
  conversation_history: Anthropic.MessageParam[];

  // Concrete persona fields
  background: string;
  expertise: string[];
  war_stories: string[];
  strong_opinions: string[];
  current_obsession: string;
  blind_spots: string[];
  communication_quirks: string[];

  // Expertise dependencies
  needs_before_contributing: string[];
  asks_for_info_from: Record<string, string>;

  private _client: Anthropic | null = null;

  // Callback for tool use events (set by world for broadcasting)
  onToolUse?: (event: { agentId: string; agentName: string; toolName: string; input: unknown }) => void;
  onToolResult?: (event: { agentId: string; agentName: string; toolName: string; result: string; isError: boolean }) => void;

  constructor(config: AgentConfig) {
    this.id = config.id || uuidv4();
    this.name = config.name;
    this.description = config.description || '';
    this.personality_traits = config.personality_traits || {};
    this.speaking_style = config.speaking_style || '';
    this.interests = config.interests || [];
    this.response_tendency = config.response_tendency ?? 0.5;
    this.temperature = config.temperature ?? 0.7;
    this.tools = config.tools || [];

    // Concrete persona fields
    this.background = config.background || '';
    this.expertise = config.expertise || [];
    this.war_stories = config.war_stories || [];
    this.strong_opinions = config.strong_opinions || [];
    this.current_obsession = config.current_obsession || '';
    this.blind_spots = config.blind_spots || [];
    this.communication_quirks = config.communication_quirks || [];

    // Expertise dependencies
    this.needs_before_contributing = config.needs_before_contributing || [];
    this.asks_for_info_from = config.asks_for_info_from || {};

    // Map model names
    const modelInput = config.model || 'claude-haiku-4-5-20251001';
    this.model = MODEL_MAP[modelInput] || modelInput;

    // Build system prompt
    this.system_prompt = config.system_prompt || this.buildDefaultPrompt();

    // Runtime state
    this.status = AgentStatus.OFFLINE;
    this.last_spoke_at = null;
    this.message_count = 0;
    this.conversation_history = [];

    // Persist to database
    db.upsertAgent({
      id: this.id,
      name: this.name,
      description: this.description,
      system_prompt: this.system_prompt,
      personality_traits: this.personality_traits,
      speaking_style: this.speaking_style,
      interests: this.interests,
      response_tendency: this.response_tendency,
      temperature: this.temperature,
      model: this.model
    });
  }

  private buildDefaultPrompt(): string {
    // Start with the core identity
    let prompt = `You are ${this.name}.\n\n${this.description}\n`;

    // Add any extra structured fields if present (backwards compatibility)
    if (this.background) {
      prompt += `\n${this.background}\n`;
    }

    if (this.speaking_style) {
      prompt += `\nSpeaking style: ${this.speaking_style}\n`;
    }

    // Add expertise and dependencies
    if (this.expertise.length > 0) {
      prompt += `\nYOU PROVIDE (your expertise - others will ask you for this):
${this.expertise.map(e => `• ${e}`).join('\n')}
`;
    }

    if (this.needs_before_contributing.length > 0) {
      prompt += `\nYOU NEED BEFORE CONTRIBUTING:
${this.needs_before_contributing.map(n => `• ${n}`).join('\n')}
`;
    }

    if (Object.keys(this.asks_for_info_from).length > 0) {
      prompt += `\nWHO TO ASK:
${Object.entries(this.asks_for_info_from).map(([info, role]) => `• "${info}" → ask the ${role}`).join('\n')}

When you need information, @mention the specific person who can provide it.
If you're missing info you need, ASK for it before proceeding.
When someone asks about your expertise, provide a clear, actionable answer.
`;
    }

    // Workspace and tools info
    prompt += `
YOUR TOOLS - Use the RIGHT tool:

1. str_replace_based_edit_tool - For CODE and structured files
   - Write actual code (scripts, config files, source code)
   - Use command: 'create' with path and file_text
   - Use command: 'view' to see existing files
   - Files persist in your workspace

2. memory - For notes and quick records
   - Store meeting notes, decisions, quick references
   - Use 'view' first to check what exists
   - Good for temporary working notes

3. bash - For running commands and getting real data
   - Run calculations, process data, check system state
   - Execute scripts you've written
   - Get current date/time, file listings, etc.

4. web_search - For current information
   - Look up facts, prices, statistics
   - Research competitors, market data
   - Find documentation or best practices

PICK THE RIGHT TOOL:
- Writing code → str_replace_based_edit_tool
- Need current info → web_search
- Need to calculate/process → bash
- Quick notes → memory

ACTION MODE - NO MORE MEETINGS

You are a DOER, not a TALKER. When given a task:

DO:
- Take immediate action using your tools
- Create deliverables (documents, analyses, code, files)
- Make decisions and state them clearly
- Use the APPROPRIATE tool for the task
- Report what you COMPLETED

DO NOT:
- Say "I'll work on X" - just DO it now
- Schedule future meetings or syncs
- Say "let's discuss" or "we should talk about"
- Write vague plans without concrete deliverables
- Defer action to later

RESPONSE FORMAT:
- 1-2 sentences max
- State what you DID or DECIDED
- If you used a tool, say what you created
`;

    return prompt;
  }

  /**
   * Initialize the Anthropic client.
   */
  async connect(): Promise<void> {
    const apiKey = process.env.ANTHROPIC_API_KEY;
    if (!apiKey) {
      throw new Error('ANTHROPIC_API_KEY environment variable not set');
    }

    this._client = new Anthropic({ apiKey });
    this.status = AgentStatus.IDLE;
    this.conversation_history = [];
    db.updateAgentStatus(this.id, 'idle');
    console.log(`Agent ${this.name} connected`);
  }

  /**
   * Disconnect the agent.
   */
  async disconnect(): Promise<void> {
    this._client = null;
    this.status = AgentStatus.OFFLINE;
    this.conversation_history = [];
    db.updateAgentStatus(this.id, 'offline');
    console.log(`Agent ${this.name} disconnected`);
  }

  /**
   * Execute one step - generate a response to the current chat context.
   * Implements agentic loop for tool execution.
   */
  async step(chatContext: string, roomId: string): Promise<string | null> {
    // Connect on first use if not already connected
    if (!this._client) {
      await this.connect();
    }

    this.status = AgentStatus.THINKING;
    console.log(`Agent ${this.name} thinking...`);

    try {
      // Build the message
      const userMessage = `Current conversation:

${chatContext}

Now respond as ${this.name}. Stay in character and keep it brief.
Use your tools if you need to take action or remember something.`;

      // Add to conversation history
      this.conversation_history.push({
        role: 'user',
        content: userMessage
      });

      // Keep history manageable (last 40 messages)
      if (this.conversation_history.length > 40) {
        this.conversation_history = this.conversation_history.slice(-40);
      }

      // Get tool definitions
      const tools = getAllToolDefinitions();

      // Build tool context
      const toolCtx: ToolContext = {
        roomId,
        agentId: this.id,
        agentName: this.name,
        workDir: getAgentWorkspace(roomId, this.id),
        sharedDir: getSharedWorkspace(roomId)
      };

      // Call Anthropic API with tools
      let response = await this._client.messages.create({
        model: this.model,
        max_tokens: 10000,
        system: this.system_prompt,
        messages: this.conversation_history,
        temperature: this.temperature,
        tools: tools as Anthropic.Tool[]
      });

      // Agentic loop - process tool calls until done
      let loopCount = 0;
      const maxLoops = 10; // Safety limit

      while (response.stop_reason === 'tool_use' && loopCount < maxLoops) {
        loopCount++;
        console.log(`Agent ${this.name} using tools (loop ${loopCount})...`);

        // Add assistant response to history
        this.conversation_history.push({
          role: 'assistant',
          content: response.content
        });

        // Process each tool use block
        const toolResults: Anthropic.ToolResultBlockParam[] = [];

        for (const block of response.content) {
          if (block.type === 'tool_use') {
            console.log(`  Tool: ${block.name}`);

            // Log tool use to database
            db.logEvent('tool_use', {
              agent_id: this.id,
              agent_name: this.name,
              tool_name: block.name,
              tool_input: block.input,
              room_id: roomId
            });

            // Emit tool use event for UI
            if (this.onToolUse) {
              this.onToolUse({
                agentId: this.id,
                agentName: this.name,
                toolName: block.name,
                input: block.input
              });
            }

            const result = await executeTool(block.name, block.input as Record<string, unknown>, toolCtx);

            // Log tool result
            db.logEvent('tool_result', {
              agent_id: this.id,
              agent_name: this.name,
              tool_name: block.name,
              is_error: result.is_error || false,
              result_length: result.content.length
            });

            // Emit tool result event for UI
            if (this.onToolResult) {
              this.onToolResult({
                agentId: this.id,
                agentName: this.name,
                toolName: block.name,
                result: result.content.slice(0, 500),  // Truncate for UI
                isError: result.is_error || false
              });
            }

            toolResults.push({
              type: 'tool_result',
              tool_use_id: block.id,
              content: result.content,
              is_error: result.is_error
            });
          }
        }

        // Add tool results to history
        this.conversation_history.push({
          role: 'user',
          content: toolResults
        });

        // Continue the conversation
        response = await this._client.messages.create({
          model: this.model,
          max_tokens: 10000,
          system: this.system_prompt,
          messages: this.conversation_history,
          temperature: this.temperature,
          tools: tools as Anthropic.Tool[]
        });
      }

      // Extract text from final response
      const textBlocks = response.content.filter(
        (block): block is Anthropic.TextBlock => block.type === 'text'
      );
      const responseText = textBlocks.map(b => b.text).join('\n').trim();

      // No response? Don't touch history, return null
      if (!responseText) {
        return null;
      }

      // Has content - add to history
      this.conversation_history.push({
        role: 'assistant',
        content: response.content
      });

      this.status = AgentStatus.SPEAKING;
      this.last_spoke_at = new Date();
      this.message_count += 1;

      console.log(`Agent ${this.name} responded: ${responseText.slice(0, 50)}...`);
      return responseText;

    } catch (error: unknown) {
      const err = error as Error & { status?: number; error?: { message?: string } };
      console.error(`Agent ${this.name} API error:`);
      console.error(`  Status: ${err.status || 'unknown'}`);
      console.error(`  Message: ${err.message || err.error?.message || 'unknown'}`);
      if (err.stack) console.error(`  Stack: ${err.stack.split('\n')[1]}`);
      this.status = AgentStatus.IDLE;
      return null;
    } finally {
      this.status = AgentStatus.IDLE;
    }
  }

  /**
   * Generate a response (alias for step).
   */
  async respond(context: string, roomId: string): Promise<string | null> {
    return this.step(context, roomId);
  }

  /**
   * Calculate probability that this agent should respond.
   */
  shouldRespond(message: Message, allAgents: string[]): number {
    let baseProb = this.response_tendency * 0.3;

    // Direct mention - high priority
    const mentionsLower = message.mentions.map(m => m.toLowerCase());
    if (mentionsLower.includes(this.name.toLowerCase())) {
      return Math.min(0.95, baseProb + 0.6);
    }

    // Question - medium boost
    if (message.content.includes('?')) {
      baseProb += 0.15;
    }

    // Topic matches interests
    const contentLower = message.content.toLowerCase();
    for (const interest of this.interests) {
      if (contentLower.includes(interest.toLowerCase())) {
        baseProb += 0.1;
        break;
      }
    }

    // Recently spoke - reduce
    if (this.last_spoke_at) {
      const secondsAgo = (Date.now() - this.last_spoke_at.getTime()) / 1000;
      if (secondsAgo < 10) {
        baseProb *= 0.3;
      } else if (secondsAgo < 30) {
        baseProb *= 0.6;
      }
    }

    return Math.min(0.8, baseProb);
  }

  /**
   * Create an agent from a config dict.
   */
  static fromConfig(config: AgentConfig): Agent {
    return new Agent(config);
  }

  /**
   * Serialize agent to dictionary.
   */
  toDict(): AgentData {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      personality_traits: this.personality_traits,
      speaking_style: this.speaking_style,
      interests: this.interests,
      response_tendency: this.response_tendency,
      temperature: this.temperature,
      model: this.model,
      status: this.status,
      message_count: this.message_count,
      last_spoke_at: this.last_spoke_at?.toISOString() || null,
      expertise: this.expertise,
      needs_before_contributing: this.needs_before_contributing,
      asks_for_info_from: this.asks_for_info_from
    };
  }
}
